### ISSUE TITLE
Reduce number of `NDArray.__init__()` overloads and make it a pure container

### ISSUE THREAD

**Issue by forFudan** [Mojo-Numerics-and-Algorithms-group/NuMojo#110]
This proposal is for long term.

## Situation

Currently, we have many overloads for `__init__()` of `NDArray` struct. On the one hand, overload is a good feature that allows more flexibility. On the other hand, it prevents developers/users conveniently checking the arguments during programming, and it means that we are using different overloading functions in other functions/modules. This causes ambiguity and difficulties in maintenance. Thus, we should try to avoid using overloads unless it is necessary.

## Solution

For `numpy`, the solution is quite straightforward: there is only one initializer, which reads in (1) shape (tuple), (2) buffer (can left empty), (3) order (C or F), etc. For other constructors (fill a value, randomize, from string), use separate construction functions.

This makes the function signature clear and makes the `ndarray` type a pure container.

I think we can also follow this. For `NDArray`, we only keep ONE initializer: `fn __init__[dtype](shape, buffer, order)`. All other constructors should be included in the module `array_creation_routines`.

## Obstacle

The biggest obstacle at the moment is that the `Tuple` type is not iterable. This means that we cannot calculate size, strides, etc, from the shape (as `Tuple`). In future, when `Tuple` becomes iterable, we can completely align the arguments with numpy.

**Comment by MadAlex1997** [comment]
Why tuple and not List or ListLiteral for shape? Tuple is not only not iterable it doesn't implement setting values either, and both Tuple and StaticTuple have sizes set at compile time (meaning the number of dimensions is set at compile time). ListLiteral also has some problems (having to do a loop to move it into a List, which I guess you can do to a Tuple as well), why is there not a trait for collections of scalar types? 

With conditional conformance now implemented we might be able to make initializers that cover two traits Takeable for the types that implement `get` but not `__get_item__` and another that covers types with `__get_item__` methods.

Another thing to consider is having an `array` function with overloads that generate `NDarray`s which would make it more similar to how most people use numpy initialization `np.array(anything_that_this_crazy_thing_can_accept)`.

**Comment by forFudan** [comment]
> Why tuple and not List or ListLiteral for shape?
List is indeed better as we sometimes need to infer shape from input (e..g, str). 

The main concern of this Issue is to reduce the number of reload of `__init__`.  Too many overload means ambiguity and difficulty in maintenance. Reducing number of reloads (options) means reducing possibility of mistakes.

Here is all the constructors we have at the moment:

1. `fn __init__( inout self, shape: List[Int], fill: Optional[Scalar[dtype]] = None, order: String = "C")`
2. `fn __init__(inout self, *shape: Int, fill: Optional[Scalar[dtype]] = None, order: String = "C")`
3. `fn __init__(inout self,  shape: VariadicList[Int], fill: Optional[Scalar[dtype]] = None, order: String = "C")`
4. `fn __init__(inout self, shape: NDArrayShape, fill: Optional[Scalar[dtype]] = None, order: String = "C")`
5. `fn __init__(inout self, data: List[SIMD[dtype, 1]], shape: List[Int], order: String = "C")`
6. `fn __init__(inout self,*shape: Int, min: Scalar[dtype], max: Scalar[dtype], order: String = "C")`
7. `fn __init__(inout self, shape: List[Int], min: Scalar[dtype], max: Scalar[dtype], order: String = "C")`
8. `fn __init__(inout self,text: String,order: String = "C",) `
9. `fn __init__(inout self,ndim: Int,offset: Int,size: Int,shape: List[Int],strides: List[Int],coefficient: List[Int],order: String = "C",)`
10. `fn __init__(inout self, data: PythonObject, order: String = "C")`

I think that:

1. This can be the only `__init__` for `NDArray`.
2. Maybe remove this. `shape` should be better in a container. Also aligned with `numpy`.
3. Similar to (1). Either keep or remove.
4. Maybe remove this, so that `shape` can never be an `NDArray`.
5. This should go to a function `numojo.array(object, shape, order)`, consistent with `numpy.
6. This should go to `random` module.
7. This should go to `random` module.
8. This should go to a function `numojo.from_string(string, order)`.
9. Many arguments of this initializer can actually be calculated from other arugments. For example, size can be calculated directly from shape. We can remove this.
10. This shoud go to a function `numojo.from_numpy`.

To summarize, we only use `List` as the type of `shape`.  Or other constructors goes to functions like `numpy.array`, `numpy.full`, `numpy.identity`...

**Comment by shivasankarka** [comment]
@MadAlex1997 @forFudan 
I agree with some of the points such as moving `6.` and `7.` to random module. 
- `9.` was meant to be only used internally, it's never meant to be exposed to the user. We can keep it or make it an internal function explicitly. 
- We can remove `3.` since it's equivalent to `2.`. 
- `5.` and `8.` should be external function like `numojo.array()`.

**Comment by forFudan** [comment]
> * `5.` and `8.` should be external function like `numojo.array()`.

Yes, `5` is more laike `numojo.array()` that reads in an data object.

`8` is not in numpy. I will create a `from_str` function in creation routine module and migrate there.

**Comment by shivasankarka** [comment]
@forFudan I think it's better to keep both `5` and `5` under `numojo.array()` and not separate it.

**Comment by forFudan** [comment]
This has been addressed by PR #139.
